lib/stripe/credit_card_cloner.rb

```diff
@@ -1,38 +1,115 @@
+# authorized (e.g. for SCA), we attach metadata { clone: true } to the card the first time we
+# clone it and look for a card with the same fingerprint (hash of the card number) and
+# that metadata key to avoid cloning it multiple times.
+
 module Stripe
   class CreditCardCloner
+    def find_or_clone(card, connected_account_id)
+      if card.user && cloned_card = find_cloned_card(card, connected_account_id)
+        cloned_card
+      else
+        clone(card, connected_account_id)
+      end
+    end
+
+    def destroy_clones(card)
+      card.user.customers.each do |customer|
+        next unless stripe_account = customer.enterprise.stripe_account&.stripe_user_id
+
+        customer_id, _payment_method_id = find_cloned_card(card, stripe_account)
+        next unless customer_id
+
+        customer = Stripe::Customer.retrieve(customer_id, stripe_account: stripe_account)
+        customer&.delete unless customer.deleted?
+      end
+    end
+
+    private
+
     def clone(credit_card, connected_account_id)
       new_payment_method = clone_payment_method(credit_card, connected_account_id)
 
       # If no customer is given, it will clone the payment method only
-      return nil, new_payment_method.id if credit_card.gateway_customer_profile_id.blank?
+      return [nil, new_payment_method.id] if credit_card.gateway_customer_profile_id.blank?
 
       new_customer = Stripe::Customer.create({ email: credit_card.user.email },
                                              stripe_account: connected_account_id)
       attach_payment_method_to_customer(new_payment_method.id,
                                         new_customer.id,
                                         connected_account_id)
 
+      add_metadata_to_payment_method(new_payment_method.id, connected_account_id)
+
       [new_customer.id, new_payment_method.id]
     end
 
-    private
+    def find_cloned_card(card, connected_account_id)
+      return nil unless fingerprint = fingerprint_for_card(card)
+
+      find_customers(card.user.email, connected_account_id).each do |customer|
+        find_payment_methods(customer.id, connected_account_id).each do |payment_method|
+          if payment_method_is_clone?(payment_method, fingerprint)
+            return [customer.id, payment_method.id]
+          end
+        end
+      end
+      nil
+    end
+
+    def payment_method_is_clone?(payment_method, fingerprint)
+      payment_method.card.fingerprint == fingerprint && payment_method.metadata["ofn-clone"]
+    end
+
+    def fingerprint_for_card(card)
+      Stripe::PaymentMethod.retrieve(card.gateway_payment_profile_id).card.fingerprint
+    end
+
+    def find_customers(email, connected_account_id)
+      start_after, customers = nil, []
+
+      (1..request_limit = 100).each do |request_number|
+        response = Stripe::Customer.list({ email: email, starting_after: start_after, limit: 100 },
+                                         stripe_account: connected_account_id)
```

Comment:
Can you explain a bit what's happening in this file and why it's needed? We're fetching some customer data but looping with up to 100 requests..?

Reply:
Ah yes. This file 😁. 

There's some additional context [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/6241#discussion_r536237764). 

The problem we're trying to solve is that there are cards stored on the platform (aka instance level) Stripe account. Prior to this SCA subscription work, we were cloning that card (and associated customer) to the shop (aka "connected" or enterprise level) Stripe account each time we needed to authorize it or take payment from it. 

On my dev environment it's resulted in dozens of cloned customers/cards. I don't know how many there might be in a production environment, but it could be a lot. 

With these changes, we no longer have to clone the card each time but instead store metadata on the cloned card so that we can go back and find it, but it means potentially sifting through lots of previously cloned cards to do so. 

However, even with these changes, our database still doesn't store any record of the cloned card, it's only at Stripe. Luis argued against adding another table and complexity to our data model (see link to context above). Both you and Maikel, however, have pointed out downsides to *not* doing so. 

It might be worth revisiting this decision, potentially even as a prerequisite to merging this. I don't mind rewriting this bit if we're all okay with the additional table.

Reply:
I think the consensus here is that we're going to keep it out of the data model and potentially revisit after the deadline.